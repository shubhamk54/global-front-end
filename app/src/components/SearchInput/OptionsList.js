import PropTypes from 'prop-types';
import React from 'react';
import './OptionsList.scss';


function OptionsList({ ...props }) {
  const { options, selectedOptionIndex, onClick } = props;

  return (
    <ul className="suggestions">
      {options.map((option, optionIndex) => (
        <li
          className={optionIndex === selectedOptionIndex ? 'suggestion-active' : ''}
          key={option}
          onClick={() => onClick(option)}
        >
          {option}
        </li>
      ))}
    </ul>
  );
}

OptionsList.propTypes = {
  options: PropTypes.array.isRequired,
  selectedOptionIndex: PropTypes.number,
  onClick: PropTypes.func,
};


export default OptionsList;
