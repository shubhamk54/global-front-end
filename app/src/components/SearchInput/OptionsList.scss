.suggestions {
  position: absolute;
  z-index: 2;
  background: white;
  border: 1px solid #999;
  border-top-width: 0;
  list-style: none;
  margin-top: 1px;
  overflow-y: auto;
  padding-left: 0;
  min-width: -webkit-fill-available;
  box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
  border-radius: 3px 3px 14px 14px;
}

.suggestions li {
  padding: 0.5rem;
  text-decoration: none;
  display: block;
}

.suggestion-active,
.suggestions li:hover {
  background-color: #ddd;
  cursor: pointer;
}

.suggestions li:not(:last-of-type) {
  border-bottom: 1px solid #999;
}
