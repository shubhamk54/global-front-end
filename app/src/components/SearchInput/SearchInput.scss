.search-input {
  position: relative;
  display: inline-block;
  input {
    padding: 6px;
    font-size: 15px;
    font-weight: 400;
    float: left;
  }
  button {
    padding: 6px;
    background: #2196f3;
    color: white;
    font-size: 17px;
    border: none;
    cursor: pointer;
  }

  button:hover {
    background: #0b7dda;
  }
}
