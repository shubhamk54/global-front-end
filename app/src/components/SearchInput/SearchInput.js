import React, { Component } from "react";
import PropTypes from "prop-types";

import OptionList from './OptionsList';

import './SearchInput.scss';

const KeyBoardBinding = {
    ENTER: 13,
    UP_ARROW: 38,
    DOWN_ARROW: 40,
}

class SearchInput extends Component {

    constructor(props) {
        super(props);

        this.state = {
            activeOption: -1,
            filteredOptions: [],
            showOptions: false,
        };
    }

    onChange = e => {
        const inputValue = e.target.value;
        const filteredOptions = this.props.suggestedOptions.filter(
            suggestion =>
                suggestion.toLowerCase().indexOf(inputValue.toLowerCase()) > -1
        );

        this.setState({
            filteredOptions,
            showOptions: true,
        }, () => this.props.onChange(this.props.name, inputValue, true));
    };

    onOptionClick = value => {
        this.setState({
            activeOption: 0,
            filteredOptions: [],
            showOptions: false,
        }, () => this.props.onChange(this.props.name, value));
    };

    onKeyDown = e => {
        const { activeOption, filteredOptions } = this.state;

        // User pressed the enter key
        if (e.keyCode === KeyBoardBinding.ENTER) {
            this.setState({
                activeOption: 0,
                showOptions: false,
            }, () => this.props.onChange(this.props.name, filteredOptions[activeOption]));
        }
        // User pressed the up arrow
        else if (e.keyCode === KeyBoardBinding.UP_ARROW) {
            if (activeOption === 0) {
                return;
            }
            this.setState({ activeOption: activeOption - 1 });
        }
        // User pressed the down arrow
        else if (e.keyCode === KeyBoardBinding.DOWN_ARROW) {
            if (activeOption - 1 === filteredOptions.length) {
                return;
            }
            this.setState({ activeOption: activeOption + 1 });
        }
    };

    onSearchClick = e => {
        this.setState({
            showOptions: false,
        }, () => this.props.onSearchClick());
    };

    onInputClick = e => {
        this.setState(prevState => ({
            showOptions: !prevState.showOptions
        }));
    }

    render() {
        return (
            <div className="search-input">
                <input
                    type="text"
                    onChange={this.onChange}
                    onKeyDown={this.onKeyDown}
                    onClick={this.onInputClick}
                    value={this.props.value}
                    placeholder={this.props.placeholder}
                />
                <button
                    onClick={this.onSearchClick}
                ><i className="fa fa-search"></i></button>

                {this.state.filteredOptions.length > 0 && this.state.showOptions && this.props.value && <OptionList
                    options={this.state.filteredOptions}
                    onClick={this.onOptionClick}
                    selectedOptionIndex={this.state.activeOption}
                />}
            </div>
        );
    }
}



SearchInput.propTypes = {
    name: PropTypes.string,
    value: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
    ]).isRequired,
    onChange: PropTypes.func.isRequired,
    suggestedOptions: PropTypes.array,
    onSearchClick: PropTypes.func,
};

SearchInput.defaultProps = {
    suggestedOptions: [],
    placeholder: 'Search...',
}

export default SearchInput;
