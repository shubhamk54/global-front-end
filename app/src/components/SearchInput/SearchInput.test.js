
import React from 'react';
import SearchInput from './SearchInput.js';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

const fakeProps = {
    suggestedOptions: [],
    name: 'test',
    value: 'test val',
    placeholder: 'Search...',
    onChange: () => { },
};

describe('Snapshot testing for SearchInput component', () => {

    it('matches the snapshot', () => {
        const tree = renderer.create(<SearchInput {...fakeProps} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

test('Should render SearchInput component', () => {
    const wrapper = mount(
        <SearchInput {...fakeProps} />
    );
    const searchInputWrapper = wrapper.find('.search-input');
    expect(searchInputWrapper.text()).toBeTruthy();
});