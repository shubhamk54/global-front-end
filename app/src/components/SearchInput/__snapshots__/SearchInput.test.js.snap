// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing for SearchInput component matches the snapshot 1`] = `
<div
  className="search-input"
>
  <input
    onChange={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    placeholder="Search..."
    type="text"
    value="test val"
  />
  <button
    onClick={[Function]}
  >
    <i
      className="fa fa-search"
    />
  </button>
</div>
`;
